File: fraction/src/egyptian_frac.py
#!/usr/bin/python3

import queue

def gcd(a, b):
    if a < b:
        a, b = b, a
    while b != 0:
        a, b = b, a % b
    return a

def frac_reduce(b, a):
    d = gcd(b, a)
    return (b // d, a // d)

def compare(xs, ys):
    if not xs:
        return ys
    elif not ys:
        return xs
    elif len(xs) < len(ys):
        return xs
    elif len(ys) < len(xs):
        return ys
    elif xs[0] < ys[0]:
        return xs
    else:
        return ys

# Decompose b/a within n Egyptian (unit) fractions (0 < b < a, irreducible).
def decompose(b, a):
    if b == 1:
        return [a]
    n, best = b, []
    cs = queue.Queue()
    cs.put((b, a, []))
    while not cs.empty():
        b, a, qs = cs.get()
        m = n - len(qs)
        low = max(a // b + 1, qs[0] + 1 if qs else 0)
        for q in range(low, a + 1):
            if best != [] and len(qs) >= len(best) - 1:
                return best
            if a*m < b*q: # m/q < b/a
                break     # b/a can not be decomposed within n fractions <= 1/q
            b1, a1 = frac_reduce(b*q - a, a*q)        # b'/a' = b/a - 1/q
            if b1 == 1 and a1 > q:
                best = compare([a1, q] + qs, best)
            else:
                cs.put((b1, a1, [q] + qs))
    return best

# Fibonacci decomposition (greedy method)
def fibonacci_decompose(b, a):
    qs = []
    while b != 1:
        q = a // b + 1
        qs.append(q)
        b, a = frac_reduce(b*q - a, a*q)
    qs.append(a)
    return qs

def verify(b, a):
    qs1 = fibonacci_decompose(b, a)
    print(f"Fibonacci decompose of {b}/{a}: {qs1}")
    qs2 = decompose(b, a)
    print(f"best decompose: {qs2}")
    b1, a1 = 0, 1
    for q in qs2:
        b1, a1 = frac_reduce(b1*q + a1, a1 * q) #b1/a1 + 1/q
    print(f"sum to: {b1}/{a1}")
    assert(b1 == b and a1 == a)
    assert(len(qs2) <= len(qs1))

if __name__ == "__main__":
    verify(1, 2)    # ==> [2]
    verify(7, 8)    # ==> [8, 4, 2]
    verify(5, 121)  # ==> [363, 121, 33]

File: fraction/src/EgyptianFrac.hs
module EgyptianFrac where

import Data.Sequence (Seq, ViewL(..), viewl, empty, (|>))
import Test.QuickCheck

reduceF b a = (b `div` d, a `div` d) where d = gcd b a

-- b/a + d/c
add (b, a) (d, c) = reduceF (c*b + a*d) (a*c)
diff (b, a) (d, c) = (b, a) `add` (-d, c)

-- sum of 1/a_1 + 1/a_2 + ... + 1/a_n
sumR as = foldr add (0, 1) [(1, a) | a <- as]

-- Choose the 'better' decomposition of Egyptian fractions
cmp [] ys = ys
cmp xs [] = xs
cmp xs ys | length xs < length ys = xs
          | length ys < length xs = ys
          | head xs < head ys = xs
          | otherwise = ys

-- Fibonacci (greedy) decomposition
-- decompose b/a = 1/q + decompose (b/a - 1/q), where q = 1 + floor(a/b)
decomposeF (1, a) = [a]
decomposeF (b, a) = q : decomposeF ((b, a) `diff` (1, q)) where q = 1 + a `div` b

decompose (1, a) = [a]
decompose (b, a) = bfs (b, a, 1 + a `div` b, [], b) [] empty where
  bfs (b, a, q, qs, n) best queue
    | best /= [] && 1 + length qs >= length best = best
    | a * n < b * q = case viewl queue of
        EmptyL -> best
        c :< cs -> bfs (from c) best cs
    | otherwise = if b' == 1 && a' > q
        then bfs (b, a, q + 1, qs, n) (cmp (a':q:qs) best) queue
        else bfs (b, a, q + 1, qs, n) best (queue |> (b', a', q:qs, n - 1))
    where
      (b', a') = (b, a) `diff` (1, q)
  from (b, a, []  , n) = (b, a, 1 + a `div` b,         []  , n)
  from (b, a, q:qs, n) = (b, a, 1 + max q (a `div` b), q:qs, n)

-- some hard cases:
-- (5, 121) (27, 29), (65, 87)

-- bound the nominator and denominator in range [1, 125]
fracOf (x, y) = reduceF (min x' y') (max x' y') where
  x' = 1 + (abs x) `mod` 125
  y' = 1 + (abs y) `mod` 125

verifySum x f = x == sumR (f x)

prop_sumF :: (Int, Int) -> Bool
prop_sumF x = verifySum (fracOf x) decomposeF

prop_sum :: (Int, Int) -> Bool
prop_sum x = verifySum (fracOf x) decompose

prop_best :: (Int, Int) -> Bool
prop_best x = qs1 == cmp qs1 qs2 where
  qs1 = decompose y
  qs2 = reverse (decomposeF y)
  y = fracOf x
